### PR TITLE
Hyper-V: Added 8 VHDx Resize tests

### PR DESCRIPTION
**Pull request by santoshx**
VHDX-RESIZE-OVER2TB-SCSI
VHDX-RESIZE-OVER2TB-IDE
VHDX-RESIZE-GROWSHRINK-SCSI
VHDX-RESIZE-GROWSHRINK-SCSI-OFFLINE
VHDX-RESIZE-GROWOVER2TB-SHRINK-SCSI-ONLINE
VHDX-RESIZE-GROWOVER2TB-SHRINK-SCSI-OFFLINE
VHDX-RESIZE-GROWSHRINK-IDE
VHDX-RESIZE-GROWOVER2TB-SHRINK-IDE

### CODE REVIEW

**Comment by santoshx** [comment]
Test Results Summary
ICA test run on 10/19/2018 19:43:49
VHD under test disco.vhdx
ARM Image under test  :  :  : 
Total Executed TestCases        2 (       2 Pass,        0 Fail,        0 Abort)
Total Execution Time(dd:hh:mm) 0:0:18
XML file: TestConfiguration

#Sr. Test Name : Test Result : Test Duration (in minutes)
----------------------------------------------------
1. VHDX-RESIZE-GROWSHRINK-SCSI : PASS : 9.16
2. VHDX-RESIZE-GROWSHRINK-SCSI-OFFLINE : PASS : 9.03

**Comment by ader1990** [comment]
the commit message is wrong, do you have 4 or 2 tcs?

File: Testscripts/Linux/STOR_VHDXResize_PartitionDiskOver2TB.sh [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        count=`expr $count + 1`
+    else
+        mkfs -t $fs "$deviceName"1
+        check_exit_status "Format partition with $fs" "LogMsg"
+        break
+    fi
+done
+
+if [ $count -eq ${#fileSystems[@]} ]; then
+    LogErr "Failed to format partition with ext4 and ext3"
```
Comment (ader1990) [inline Testscripts/Linux/STOR_VHDXResize_PartitionDiskOver2TB.sh]: you cannot format a partition with 2 formats at the same time

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+{
+	param ($vhdpath, $newsize, $controllertype, $vmname, $hvserver, $ip, $port)
+
+	# for IDE & offline need to stop VM before resize
+	if ( $controllertype -eq "IDE" -or $testParameters.Offline -eq "True") {
+		LogMsg "Stopping VM for IDE disk or offline is requested"
+		Stop-VM -VMName $vmname -ComputerName $hvserver -force
+	}
+
+	$newvhdxsize = Convert-StringToUInt64 $newsize
```
Comment (ader1990) [inline Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: use camel casing for variables in powershell
$newVhdxSize

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	Resize-VHD -Path $vhdpath -SizeBytes ($newvhdxsize) -ComputerName $hvserver
+	if (-not $?) {
+		$testResult = "FAIL"
+		Throw "Unable to grow VHDX file ${vhdpath}"
+	}
+
+	LogMsg "Let system have some time for the volume change to be indicated. Sleep 5 ..."
+	Start-Sleep -s 5
```
Comment (ader1990) [inline Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: use a retry/check pattern here like Wait-VmState

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+		if (-not (Wait-ForVMToStartKVP $vmname $hvserver $timeout )) {
+			$testResult = "FAIL"
+			Throw "${vmname} failed to start"
+		} else {
+			LogMsg "Started VM ${vmname}"
+		}
+	}
+
+	# check file size after resize
+	$vhdxinforesize = Get-VHD -Path $vhdpath -ComputerName $hvserver
```
Comment (ader1990) [inline Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: ditto

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+	if ($disksize -ne $newvhdxsize) {
+		$testResult = "FAIL"
+		Throw "VM ${vmname} detects a disk size of ${disksize}, not the expected size of ${newvhdxsize}"
+	}
+
+	# Make sure if we can perform Read/Write operations on the guest VM
+	# if file size larger than 2T (2048G), use parted to format disk
+	if ([int]($newsize/1gb) -gt 2048) {
+		$ret = RunLinuxCmd -ip $ip -port $port -username $user -password $password -command "cp -f /home/$user/STOR_VHDXResize_PartitionDiskOver2TB.sh /root/" -runAsSudo
+		$guest_script = "STOR_VHDXResize_PartitionDiskOver2TB.sh"
```
Comment (ader1990) [inline Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: use camelCasing for PS variables -- this applies for all the future/past variables in your PowerShell Code

**Comment by santoshx** [comment]
> the commit message is wrong, do you have 4 or 2 tcs?

Commit message was corrected already.

File: Testscripts/Linux/STOR_VHDXResize_PartitionDiskOver2TB.sh (hunk 1 above)
Comment (santoshx) [reply Testscripts/Linux/STOR_VHDXResize_PartitionDiskOver2TB.sh]: Here we are not formatting partitions. We stating that format was failed for ext4 and ext 3. 
However I changed it to be more clear.

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 (hunk 2 above)
Comment (santoshx) [reply Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: Done

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 (hunk 4 above)
Comment (santoshx) [reply Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: done

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 (hunk 3 above)
Comment (santoshx) [reply Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: Here we have changed the size of vhdx outside of VM. VM state won't help. This wait is for Windows filesystem to stabilize.

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 (hunk 5 above)
Comment (santoshx) [reply Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: Done

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 (hunk 3 above)
Comment (santoshx) [reply Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: Thank you for pushing me to remove the Sleep.
I removed it and it doesn't have any impact on success of tests.

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 [hunk 6]
```diff
@@ -0,0 +1,199 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: @santoshx what option do you have for indentation? tab or space? what is the indentation size?

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+.Parameter testParams
+    Test data for this test case
+#>
+
+Param([String] $TestParams)
+
+$ErrorActionPreference = "Stop"
+$testResult = "FAIL"
+
+Function ResizeVerifyHardDisk
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: Change to Set-HardDiskSize

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+			$newSize = $testParameters.shrinkSize
+			ResizeVerifyHardDisk $vhdPath $newSize $controllerType $vmName $hvServer $ip $port
+		}
+		if ($null -ne $testParameters.newSize) {
+			$newSize = $testParameters.newSize
+			ResizeVerifyHardDisk $vhdPath $newSize $controllerType $vmName $hvServer $ip $port
+		}
+		$testResult = "PASS"
+
+	} catch {
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: please add exception description and its line. Find the example from existing scripts.

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 (hunk 6 above)
Comment (santoshx) [reply Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: I've tab only indentation. Tabsize is set to 4 in my vim config.

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 (hunk 8 above)
Comment (santoshx) [reply Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: Done

File: Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1 (hunk 7 above)
Comment (santoshx) [reply Testscripts/Windows/STOR-VHDX-RESIZE-GROWSHRINK.ps1]: Done

**Comment by santoshx** [comment]
@ader1990 Do you have any further comments? If not, please approve and merge it. Thanks.

**Comment by santoshx** [comment]
Test Results Summary
ICA test run on 10/24/2018 09:04:03
VHD under test disco.vhdx
ARM Image under test  :  :  : 
Total Executed TestCases        8 (       8 Pass,        0 Fail,        0 Abort)
Total Execution Time(dd:hh:mm) 0:1:46
XML file: TestConfiguration

#Sr. Test Name : Test Result : Test Duration (in minutes)
----------------------------------------------------
1. VHDX-RESIZE-GROWSHRINK-SCSI : PASS : 11.76
2. VHDX-RESIZE-GROWSHRINK-SCSI-OFFLINE : PASS : 9.45
3. VHDX-RESIZE-OVER2TB-SCSI : PASS : 14.99
4. VHDX-RESIZE-OVER2TB-IDE : PASS : 14.4
5. VHDX-RESIZE-GROWOVER2TB-SHRINK-SCSI-ONLINE : PASS : 12.87
6. VHDX-RESIZE-GROWOVER2TB-SHRINK-SCSI-OFFLINE : PASS : 15.87
7. VHDX-RESIZE-GROWSHRINK-IDE : PASS : 10.38
8. VHDX-RESIZE-GROWOVER2TB-SHRINK-IDE : PASS : 16.55

**Comment by santoshx** [comment]
@juhlee-microsoft Please merge in case further comments.

**Comment by santoshx** [comment]
Any more comments?

File: XML/TestCases/FunctionalTests.xml [hunk 9]
```diff
@@ -1370,4 +1370,140 @@
...
+		<testName>VHDX-RESIZE-GROWSHRINK-IDE</testName>
+		<setupScript>.\Testscripts\Windows\AddVhdxHardDisk.ps1</setupScript>
+		<testScript>STOR-VHDX-RESIZE-GROWSHRINK.ps1</testScript>
+		<files>.\Testscripts\Linux\utils.sh,.\Testscripts\Linux\STOR_VHDXResize_ReadWrite.sh,.\Testscripts\Linux\STOR_VHDXResize_PartitionDisk.sh,.\Testscripts\Linux\check_traces.sh</files>
+		<setupType>SingleVM</setupType>
+		<testParameters>
+			<param>IDE=0,1,Dynamic,512,3GB</param>
+			<param>growSize=12GB</param>
+			<param>shrinkSize=10GB</param>
+			<param>fileSystems=(ext4 ext3 xfs)</param>
```
[the remark this answers is not in the document]
Comment (santoshx) [reply XML/TestCases/FunctionalTests.xml]: done

File: XML/TestCases/FunctionalTests.xml [hunk 10]
```diff
@@ -1370,4 +1370,140 @@
...
+		<testName>VHDX-RESIZE-GROWSHRINK-SCSI-OFFLINE</testName>
+		<setupScript>.\Testscripts\Windows\AddVhdxHardDisk.ps1</setupScript>
+		<testScript>STOR-VHDX-RESIZE-GROWSHRINK.ps1</testScript>
+		<files>.\Testscripts\Linux\utils.sh,.\Testscripts\Linux\STOR_VHDXResize_ReadWrite.sh,.\Testscripts\Linux\STOR_VHDXResize_PartitionDisk.sh,.\Testscripts\Linux\check_traces.sh</files>
+		<setupType>SingleVM</setupType>
+		<testParameters>
+			<param>SCSI=0,1,Dynamic,512,3GB</param>
+			<param>growSize=4GB</param>
+			<param>shrinkSize=3GB</param>
+			<param>fileSystems=(ext4 ext3 xfs)</param>
```
[the remark this answers is not in the document]
Comment (santoshx) [reply XML/TestCases/FunctionalTests.xml]: done

File: XML/TestCases/FunctionalTests.xml [hunk 11]
```diff
@@ -1370,4 +1370,140 @@
...
+		<testName>VHDX-RESIZE-GROWSHRINK-SCSI</testName>
+		<setupScript>.\Testscripts\Windows\AddVhdxHardDisk.ps1</setupScript>
+		<testScript>STOR-VHDX-RESIZE-GROWSHRINK.ps1</testScript>
+		<files>.\Testscripts\Linux\utils.sh,.\Testscripts\Linux\STOR_VHDXResize_ReadWrite.sh,.\Testscripts\Linux\STOR_VHDXResize_PartitionDisk.sh,.\Testscripts\Linux\check_traces.sh</files>
+		<setupType>SingleVM</setupType>
+		<testParameters>
+			<param>SCSI=0,1,Dynamic,512,3GB</param>
+			<param>growSize=4GB</param>
+			<param>shrinkSize=3GB</param>
+			<param>fileSystems=(ext4 ext3 xfs)</param>
```
[the remark this answers is not in the document]
Comment (santoshx) [reply XML/TestCases/FunctionalTests.xml]: done

**Pull request merged**